Disable cascading writes in ambiguous cases

dev2718 opened the pull request:
## Proposed Changes
This PR serves, among others, to implement https://github.com/ElektraInitiative/libelektra/issues/3742 and https://github.com/ElektraInitiative/libelektra/issues/4012.

List of proposed changes:
* `kdb set`, `kdb meta-set`: (Resolve https://github.com/ElektraInitiative/libelektra/issues/401) Only allow writes to the cascading namespace if the lookup succeeds (Otherwise, the operation is ambiguous and therefore aborted). 
No more guessing of namespaces in case a cascading key is given (`user:, system`: for `kdb set`, `spec:` for `kdb meta-set`),
* `kdb file`: Remove namespace guessing
*  `kdb editor/import`: Disable the use of cascading names (and the 'validate' strategy operating on cascading keys) entirely to resolve https://github.com/ElektraInitiative/libelektra/issues/2762
* `kdb export` keep behaviour to not break `tests/shell/shell_recorder/shell_recorder.sh` 
* `kdb get`, `kdb sget`, `kdb meta-get`, `kdb set`, `kdb meta-set`: Use a cascading key for `parentKey` to solve https://github.com/ElektraInitiative/libelektra/issues/2561.
Introduce `-f` flag to `kdb set`, `kdb meta-set` to use old behaviour (bypass validation)
* Implement https://github.com/ElektraInitiative/libelektra/issues/4012
* Update numerous tests to comply with changes.
* Update documentation of `kdb` manpages.
* Add new subsection on cascading writes to `elektra-cascading.md`.

## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildservers are happy. If not, fix **in this order**:
  - [x] add a line in `doc/news/_preparation_next_release.md`
  - [x] reformat the code with `scripts/dev/reformat-all`
  - [x] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [x] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

mpranj commented:
Thank you so much for working on this!

Is the plan here to fully implement the proposal from #3742?

markus2330 commented:
Thank you for creating the PR!

Did you already consider that cascading writes are okay if `-N` is given? Please also update docu&test cases.

dev2718 commented:
As of ff85a90e1c15bca993df9d969fd3b847bc7e3561 `-N` is considered.
The old behaviour of guessing the flag `-N` if not supplied as either "user:" or "system:" is removed in this commit.

`Cmdline::createKey` will now translate a cascading name to a "namespaced" name if the namespace flag is provided.

All code segments in kdb using the flag have been updated:
* `kdb editor/import`: abort when `-N` is needed but not supplied as guessing was removed
* `kdb file`: removed behaviour now implemented in `Cmdline::createKey`

I plan to update documentation and testing once we have settled on the new behaviour.

markus2330 commented on src/tools/kdb/editor.cpp:
```diff
@@ -96,6 +96,12 @@ int EditorCommand::execute (Cmdline const & cl)
 	{
+		if (cl.ns.empty ())
+		{
+			cerr << "Namespace (-N) needs to be specified for strategy 'validate'." << endl;
```
For the editor for every strategy a namespace must be given explicitly.

dev2718 replied:
I made this distinction because in libelektra/doc/help/kdb-editor.md it is described:
```-N, --namespace=: Specify the namespace to use when writing cascading keys (validation strategy only).```

dev2718 replied:
Obsolete due to https://github.com/ElektraInitiative/libelektra/issues/4012.

markus2330 commented on src/tools/kdb/import.cpp:
```diff
@@ -75,6 +75,12 @@ int ImportCommand::execute (Cmdline const & cl)
 
 	if (cl.strategy == "validate")
 	{
+		if (cl.ns.empty ())
```
Also here it is not specific to validate: `kdb import` does not have a way to import without namespace.

dev2718 replied:
Same reasoning as with `kdb editor` (`/doc/help/kdb-import.md`).

markus2330 replied:
I think the docu is wrong, at least for import I am quite sure cascading imports never work (regardless of strategy). It is more safe if we disable it (for import+editor) and reintroduce it later if there is proper support for it. (unless you tested it and it works)

Can you please also add tests in which ambiguous cases are triggered (and `kdb` fails).

dev2718 replied:
Both are disabled as of now; tests with such cases are now found in `cascading.md`.

markus2330 reviewed (changes requested):
Thank you for the improvements! Docu&test cases are still missing which makes it hard to know what the intended behavior is.

markus2330 commented:
I am afraid something went wrong in the rebase. One way to resolve it is to create a new branch, cherry-pick the commits and create a new PR.

dev2718 commented:
> I am afraid something went wrong in the rebase. One way to resolve it is to create a new branch, cherry-pick the commits and create a new PR.

The rebase is fixed now.

markus2330 commented on doc/help/kdb-editor.md:
```diff
@@ -74,9 +74,7 @@ The other strategies are implemented by the merge framework and are documented i
-  Specifies which default namespace should be used when setting a cascading name.
-  By default the namespace is user, except `kdb` is used as root, then `system`
-  is the default (`validate` strategy only).
+  Specifies which default namespace should be used when setting a cascading name (`validate` strategy only).
```
Behavior on other namespaces? (Error?)

dev2718 replied:
Obsolete since https://github.com/ElektraInitiative/libelektra/issues/4012 is now implemented.

markus2330 commented on doc/help/kdb-rm.md:
```diff
@@ -11,6 +11,10 @@ Note that when using the `-r` flag, not only the key directly at `path` will be
 
+## LIMITATIONS
+
+Cascading keys (i.e. keys that start with `/`) are only allowed using the `-f` flag.
```
Why? And what is the behavior? I would expect it removes any key in any namespace. What is ambiguous here?

I would also avoid giving -f so different meanings: bypassing spec: validation (which also makes sense for kdb-rm) and allowing cascading writes. What if someone does not want to bypass validation but would like to use cascading?

dev2718 replied:
My rationale was to implement https://github.com/ElektraInitiative/libelektra/issues/3742 (I interpreted this to mean that all cascading operations on `kdb rm` were meant to be disabled, as not otherwise specified).
The -f option was intended to keep the old behavior available, but of course this is not an optimal solution.
I will revert it back.

markus2330 replied:
Yes, I think in non-ambiguous cases we can simply remove the key(s). `-f` should be only for disabling the spec validation.

The remaining problem is that repeated executions of `kdb rm -r /something` might remove several keys. Please think about it if we really want the same semantics as used in `kdb set`. How is the behavior in FUSE when you remove a file that is "backed up" by several keys (in several namespaces)?

dev2718 replied:
The old behavior is back.

Although it is not an advisable strategy to do that, I think it is better to not introduce a new special case here. and keep things consistent with the other commands.

In the FUSE tool this case does not occur, since the cascading namespace is mounted read-only (This was decided because a file system that only supports writing partially (due to the ambiguity of creating keys in the cascading namespace) is a mess regarding compatibility with tools and usability for users).

markus2330 commented on doc/help/kdb-import.md:
```diff
@@ -64,9 +64,7 @@ The other strategies are implemented by the merge framework and are documented i
-  Specifies which default namespace should be used when setting a cascading name.
-  By default the namespace is user, except `kdb` is used as root, then `system`
-  is the default (`validate` strategy only).
+  Specifies which default namespace should be used when setting a cascading name (`validate` strategy only).
```
Again: what is with other strategies?

dev2718 replied:
Without the `-N` option the validate strategy no longer makes sense; this I removed it (and updated this documentation).

markus2330 commented on doc/help/kdb-meta-get.md:
```diff
@@ -42,6 +42,8 @@ This command will return the following values as an exit status:<br>
 - `-d`, `--debug`:
   Give debug information. Prints additional debug information in case of errors/warnings.
+- `-N`, `--namespace=NS`:
+  Use the specified namespace in case the provided key does not already have a namespace.
```
What if the namespace is provided twice? (I would expect an error from this docu.)

dev2718 replied:
For a case like `kdb meta-get -N system user:/key metakey` I opted to use the namespace of the key, since this is how `kdb set` worked in the past, and throwing an error (while still being consistent with `kdb set`) leads to e.g. testcase [check_distribution.sh](https://github.com/ElektraInitiative/libelektra/blob/f910d2345fda33ce269b28f07818ed840b8881de/tests/shell/check_distribution.sh) breaking.

What do you think is the solution?

markus2330 replied:
The proper solution probably is to completely get rid of `-N` as we actually want it to be synonymous to add the namespace to the key. I proposed this in #4012

dev2718 replied:
The proposal is now fully implemented.

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -129,6 +129,7 @@ _(Michael Tucek)_
 
 - Really add all tools when using `-DTOOLS=ALL`. _(Markus Raab)_
 - ZeroMQ Hub: fix compilation and man page. _(Markus Raab)_
+- ambiguous write operations now disabled in kdb _(Alexander Firbas)_
```
Please give more details here (e.g. that it is about cascading keys, also link to the tutorial.), you can even make a highlight out of this.

dev2718 replied:
I have updated the release notes with further details.

markus2330 commented on doc/tutorials/cascading.md:
```diff
@@ -91,6 +91,43 @@ The **proc** namespace is not accessible by the command line tool **kdb**, as it
 
+## Write Operations and the cascading Namespace
+
+When a value is to be written to a cascading key, i.e. a key starting with '/', only cascading keys that resolve to some existing key may be used.
```
```suggestion
If a value is to be written to a cascading key, i.e., a key starting with '/', only cascading keys that resolve to an existing key will be used.
```

markus2330 commented on doc/tutorials/cmerge.md:
````diff
@@ -209,6 +209,12 @@ kdb get user:/tests/arrays/result/#0
+## Cleanup
+
+```sh
+kdb rm -rf user:/tests
````
Why -f if there is a namespace specified?

dev2718 replied:
That was a mistake, I removed it.

markus2330 reviewed (commented):
This PR is starting to look great! :rocket:

markus2330 commented on doc/help/kdb-dirkey.md:
```diff
@@ -0,0 +1,43 @@
+# kdb-dirkey(1) -- Get the directory name of a key name including namespace
```
Do we need this command?

dev2718 replied:
I have removed it.

markus2330 commented on doc/help/kdb-meta-set.md:
```diff
@@ -34,6 +33,8 @@ that is the place where you usually want to set metadata.
 - `-q`, `--quiet`:
   Suppress non-error messages.
+- `-f`, `--force`:
+  Do not perform a cascasing lookup if the key provided has a namespace. For example, this bypasses validation specified in the spec: namespace for the given key.
```
It is not the cascading lookup which bypasses the spec: validation.

```suggestion
  Do not perform a cascading KDB operations if the key provided has a namespace. For example, this bypasses validation specified in the spec: namespace for the given key.
```

markus2330 commented on doc/help/kdb-namespace.md:
```diff
@@ -0,0 +1,44 @@
+
+- `-n`, `--no-newline`:
+  Suppress the newline at the end of the output.
+- `-S`, `--no-separators`:
```
Is not needed, see https://github.com/ElektraInitiative/libelektra/issues/4012#issuecomment-909515824

dev2718 replied:
I have removed it.

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -33,6 +33,7 @@ docker run -it elektra/elektra
 
 ## Highlights
 
+- `kdb` no longer permits write operations on cascading keys that do not correspond to an existsing key. See the details in the `Tools` section below and the new subsection on cascading writes in the [tutorial](/doc/tutorials/cascading.md) on cascading keys for further information. _(Alexander Firbas)_
```
double negation

```suggestion
- `kdb` now prohibits write operations on cascading keys that miss a corresponding existing key. See the details in the `Tools` section below and the new subsection on cascading writes in the [tutorial](/doc/tutorials/cascading.md) on cascading keys for further information. _(Alexander Firbas)_
```

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -166,6 +167,19 @@ _(Michael Tucek)_
+  No more guessing of namespaces in case a cascading key is given (`user:, system`: for `kdb set`, `spec:` for `kdb meta-set`), _(Alexander Firbas)_
+- `kdb set`, `kdb meta-set`: Validation of keys can no longer be bypassed by using non-cascading keys (except with the new --force (-f) option). _(Alexander Firbas)_
+- Disable `-N/--namespace` option in all kdb subcommands _(Alexander Firbas)_
+- Implement new name part getter commands `kdb namespace`, `kdb basename`, `kdb dirname` and `kdb dirkey`. _(Alexander Firbas)_
```
might need update (dirkey not needed?) Please add other 3 names to doc/help/elektra-glossary.md

dev2718 replied:
The releasenotes got updated; I only added dirname as basename and namespace are already described in that file.

markus2330 commented on src/tools/kdb/nameparthelper.hpp:
```diff
@@ -0,0 +1,26 @@
+/**
```
Very good addition in the lib!

markus2330 commented on src/tools/kdb/nameparthelper.hpp:
```diff
@@ -0,0 +1,26 @@
+string keyNamespaceWithoutSeparator (Key const & key);
+string keyBasename (Key const & key);
+string keyDirname (Key const & key);
+string keyDirkey (Key const & key);
```
Is this needed?

dev2718 replied:
It now is removed.

markus2330 reviewed (changes requested):
Excellent work! Only minor things.

mpranj commented:
The manpages are not fully re-generated. The simplest is to apply the patch from the failed CI check: https://cirrus-ci.com/task/5640107244912640

dev2718 commented:
> The manpages are not fully re-generated. The simplest is to apply the patch from the failed CI check: https://cirrus-ci.com/task/5640107244912640

Thank you for the hint; Done.

mpranj commented on doc/help/kdb-meta-set.md:
```diff
@@ -34,6 +33,8 @@ that is the place where you usually want to set metadata.
 - `-q`, `--quiet`:
   Suppress non-error messages.
+- `-f`, `--force`:
+  Do not perform a cascading KDB operations if the key provided has a namespace. For example, this bypasses validation specified in the spec: namespace for the given key.
```
```suggestion
  Do not perform cascading KDB operations if the key provided has a namespace. For example, this bypasses validation specified in the spec: namespace for the given key.
```

mpranj commented on doc/help/kdb-set.md:
```diff
@@ -30,15 +31,14 @@ To set a key to a negative value, `--` has to be used to stop option processing.
 - `-d`, `--debug`:
   Give debug information. Prints additional debug information in case of errors/warnings.
+- `-f`, `--force`:
+  Do not perform a cascasing lookup if the key provided has a namespace. For example, this bypasses validation specified in the spec: namespace for the given key.
```
```suggestion
  Do not perform a cascading lookup if the key provided has a namespace. For example, this bypasses validation specified in the spec: namespace for the given key.
```

mpranj reviewed (approved):
Awesome work! :rocket:

mpranj commented:
Thank you!
